### DISCUSSION TITLE
feat: direct llama.cpp integration

### DISCUSSION THREAD

**Discussion by tjbck** [open-webui/open-webui#7543] [Ideas]

**Comment by jukofyork** [comment]
Just a quick follow-up to say it seems to work fine:

- I had to change the llama.cpp server port to 8081 to not clash with OpenWebUI (eg: `./server --port 8081 ...`).
- Then set the OpenAPI base URL to `http://127.0.0.1:8081/v1` and the API Key to not be blank (eg: `none`) in OpenWebUI settings.

and it seems to be calling the [OAI-like API](https://github.com/openai/openai-openapi) endpoint on the llama.cpp server fine. It wasn't that clear I needed to add the `/v1` to the URL and ensure the API Key not be blank though (had to find by trial and error).

The only difference I can see is there is no little "information" icon like there was with Ollama models, but it does seem to be calling the `OAI-like API` endpoint to get these stats:

```
{
  "tid": "140627543928832",
  "timestamp": 1712766280,
  "level": "INFO",
  "function": "print_timings",
  "line": 313,
  "msg": "prompt eval time     =     129.89 ms /    55 tokens (    2.36 ms per token,   423.43 tokens per second)",
  "id_slot": 0,
  "id_task": 13,
  "t_prompt_processing": 129.892,
  "n_prompt_tokens_processed": 55,
  "t_token": 2.3616727272727274,
  "n_tokens_second": 423.42869460782805
}
```

I'll report back if I can see any other major differences, but otherwise :+1:

**Reply by cjccjj** [reply]
I use the same config, added api_key and using port 8081, but Open webui refuse to work. 

Log from Open webui as blow
`open-webui  |     raise last_exc
open-webui  |           └ ClientConnectorError(ConnectionKey(host='127.0.0.1', port=8081, is_ssl=True, ssl=True, proxy=None, proxy_auth=None, proxy_hea...'
` aiohttp.client_exceptions.ClientConnectorError: Cannot connect to host 127.0.0.1:8081 ssl:default [Connect call failed ('127.0.0.1', 8081)]`

is it a SSL issue that you must use https here? can't find an env variable to bypass it. Did you use any setting to allow non ssl connection?

I can confirm the llama.cpp server is up and running well, can chat with it using llama server webui, also can read from /v1 well. Just open webui could not connect to it. Any idea?

**Reply by ubergarm** [reply]
@cjccjj 

It works fine, scroll down to see working llama.cpp and open-webui commands:

https://github.com/open-webui/open-webui/discussions/7543#discussioncomment-12229152

I don't use any SSL stuff just use `export OPENAI_API_BASE_URL="http://127.0.0.1:8080/v1"` as http without the `s` doesn't need it and is fine for local use.

**Reply by cjccjj** [reply]
Thank you. My setup is running on a VPS in docker so maybe it is different.

I somehow made it work, by putting both open-webui and llama.cpp server, in the same docker compose file. That way I use 
`http://llamacpp:8080/v1` and it worked. (llamacpp is the container name, use its ip of 172.x.x.x also works)

My previous config of 127.0.0.1 was not working becasue I forgot I was running in docker. However, host.docker.internal or 172.17.0.1 were not working neither. I have tested them in terminal, inside container, network works well for both hostname and internal ip. I have no idea why they don't work, and have to combine the docker compose to a single one, and they worked.

**Reply by ubergarm** [reply]
@cjccjj 

Probably you need to use `0.0.0.0` so the process will bind to all the network devices and not just local host `127.0.0.1`. Otherwise you can't access it from outside the docker container. Glad you got it figured out by binding to its actual IP address, though quad 0 is probably easier and more common convention. Cheers!

**Comment by jukofyork** [comment]
I've used this quite a bit with llama.cpp server now and the only problem I've come across is pressing the stop button doesn't actually disconnect/stop the generation. This was a problem with the Ollama server and was fixed AFAIK:

https://github.com/open-webui/open-webui/issues/1166
https://github.com/open-webui/open-webui/issues/1170

It would be helpful if this could be added to the OpenAI API code too, as otherwise the only way currently to stop runaway LLMs  is to Control-C the running server and restart it.

**Comment by jukofyork** [comment]
Another thing that might be helpful would be to add an option to hide the "Modelfiles" and "Prompts" menu options in the left, as these aren't able to be used with the OpenAI API and just add clutter.

**Comment by tjbck** [comment]
@jukofyork I'll start working on this feature after #665, we should strive to keep all the core features.

**Comment by DenisSergeevitch** [comment]
Small update: Stop generation button is still an issue

**Comment by justinh-rahb** [comment]
@DenisSergeevitch that is unrelated to the issue being discussed here. Let's keep discussion of the _stop generation_ function here:
- #1568

**Comment by tjbck** [comment]
Related: https://github.com/open-webui/open-webui/issues/1166

> I'm sorry, looks like it was my mistake or something with my setup (reverse proxies?) caused a problem. Can confirm that everything works as expected with the current open webui and ollama docker. :) thanks for the great software

#1568

> I've completely ditched Ollama and just moved over to the llama.cpp server and just want to say thanks as it's working really smoothly with Open-WebUI! 👍

@jukofyork @DenisSergeevitch @SN4K3D @0x7CFE

Correct me if I'm wrong but, stop generation button not actually stopping is only an issue when running LLMs with Ollama using CPU only and a vast majority of us face zero issue with terminating the response using a stop button. Could anyone confirm this with the latest? I appreciate it!

**Comment by SN4K3D** [comment]
I confirm the issue it is with ollama.cpp using LLMs CPU only, today i have try with the latest version and the stop button work , this is top all thread ollama can be launch. 
Thanks all for your work its appreciated

**Comment by jianlinshi** [comment]
I got llama.cpp server on, and set openwebui connection to http://127.0.0.1:8081/v1, however, it doesn't show any model loaded, and keep saying no model selected. 
How do you configure the arguments to make they work together?

**Reply by ubergarm** [reply]
same thing for me.. when i access llama-server via `litellm` i can put in any old `openai/foo/bar` for the model and it just works. trying to force open-webui to stop downloading random onnx stuff and just use the endpoint i'm already running lol

**Comment by ubergarm** [comment]
Okay I finally got open-webui to work with llama-server with DeepSeek-R1 671B unsloth quant.

## Start llama-server
```bash
# adjust num-gpu-layers and threads as needed
./build/bin/llama-server \
    --model "/mnt/raid/models/unsloth/DeepSeek-R1-GGUF/DeepSeek-R1-UD-Q2_K_XL/DeepSeek-R1-UD-Q2_K_XL-00001-of-00005.gguf" \
    --ctx-size 8192 \
    --n-gpu-layers 5 \
    --cache-type-k q4_0 \
    --cache-type-v f16 \
    --parallel 1 \
    --threads 16 \
    --host 127.0.0.1 \
    --port 8080
```

## install openwebui
```bash
uv venv ./venv --python 3.11
source ./venv/bin/activate
uv pip install open-webui
```

## run open-webui
This is a sane set of defaults with data in the current directoryso you can delete easily when things get borked.

Copy paste this into a script e.g. `go.sh` and `chmod u+x go.sh` then `./go.sh` and bingo
```bash
#!/usr/bin/env bash

source venv/bin/activate

# IT DOES NOT HONOR HOST and PORT ENV VAR SO PASS IT MANUALLY
# https://docs.openwebui.com/getting-started/env-configuration/#port

export DATA_DIR="$(pwd)/data"
export ENABLE_OLLAMA_API=False
export ENABLE_OPENAI_API=True
export OPENAI_API_KEY="your_secret_key"
export OPENAI_API_BASE_URL="http://127.0.0.1:8080/v1"
#export DEFAULT_MODELS="openai/foo/bar"
export WEBUI_AUTH=False
export DEFAULT_USER_ROLE="admin"
export HOST=127.0.0.1
export PORT=3000

open-webui serve \
  --host $HOST \
  --port $PORT
```

**Reply by charleswg** [reply]
Thanks for the instruction. Are you able to enable search function?

**Reply by ubergarm** [reply]
I don't even know what search function means, sorry! I have a english/chinese guide to run [DeepSeek-R1 GGUF on ktransformers](https://github.com/ubergarm/r1-ktransformers-guide/) if you are interested though. lol...

Some people are having problems with open-webui with R1 though I think because maybe open-webui does not strip the `<think></think>` tags, do you know?

https://github.com/kvcache-ai/ktransformers/issues/647

**Reply by ubergarm** [reply]
I assume you just enable web search function as normal? I dunno, I haven't tried it, but I assume it would work with R1?

Also, I got some tips here to disable title generation, tag generation, and auto-completion. Also that the think tags are removed but might have issues if you are putting `<think></think>` into the prompt directly.

https://github.com/kvcache-ai/ktransformers/issues/618#issuecomment-2681381587

**Reply by ubergarm** [reply]
> Thanks for the instruction. Are you able to enable search function?

I just tried today, web search works fine with this method. Just turn it on in the UI e.g.

Settings -> Admin Options -> Web Search -> Enable -> DuckDuckGo (or whatever you want)

Then click the `WebSearch` toggle and ask your questions

**Reply by yhyu13** [reply]
Note the loca url has to be 127.0.0.1 or 0.0.0.0 instead of localhost
